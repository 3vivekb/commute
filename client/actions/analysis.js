import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  newAnalysis.trips = []
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organzationId) => [
  deleteLocally({ analysisId: id, organzationId }),
  push(`/organizations/${organzationId}`)
]
