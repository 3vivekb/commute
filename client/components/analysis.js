import React, {Component} from 'react'

export default class Analysis extends Component {
  static propTypes = {}

  render () {
    return <h1>Analysis</h1>
  }
}
