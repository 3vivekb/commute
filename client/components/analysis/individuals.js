import React, {Component} from 'react'

export default class Individuals extends Component {
  static propTypes = {}

  render () {
    return <h1>Analysis</h1>
  }
}
