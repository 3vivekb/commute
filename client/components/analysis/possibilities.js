import React, {Component} from 'react'

export default class Possibilities extends Component {
  static propTypes = {}

  render () {
    return <h1>Analysis</h1>
  }
}
