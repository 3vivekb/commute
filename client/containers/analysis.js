import {connect} from 'react-redux'

import analysisActions from '../actions/analysis'
import _histogram from '../components/analysis/histogram'
import _individuals from '../components/analysis/individuals'
import _possibilities from '../components/analysis/possibilities'
import _summary from '../components/analysis/summary'

function mapStateToProps (state, props) {
  const {analysis, commuter, group, site} = state
  const currentAnalysis = analysis[props.params.analysisId]
  return {
    analysis: currentAnalysis,
    commuterStore: commuter,
    groupName: group[currentAnalysis.groupId].name,
    siteName: site[currentAnalysis.siteId].name
  }
}

function mapDispatchToProps (dispatch) {
  return {
    deleteAnalysis: (opts) => dispatch(analysisActions.delete(opts)),
    loadAnalysis: (opts) => dispatch(analysisActions.loadOne(opts))
  }
}

export const Histogram = connect(mapStateToProps)(_histogram)
export const Individuals = connect(mapStateToProps)(_individuals)
export const Possibilities = connect(mapStateToProps)(_possibilities)
export const Summary = connect(mapStateToProps, mapDispatchToProps)(_summary)
