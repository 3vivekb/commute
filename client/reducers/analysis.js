import update from 'react-addons-update'

import analysisDefaults from '../utils/analysisDefaults'
import {fixedRound} from '../utils/common'
import {addToEntityMap} from '../utils/entities'
import {makeGenericReducerHandlers} from '../utils/reducers'

const METRICS = Object.keys(analysisDefaults.metrics)

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'delete', 'set many'],
  name: {
    singular: 'analysis',
    plural: 'analyses'
  }
})

reducers['receive mock calculated trips'] = function (state, action) {
  const {analysisId, trips} = action.payload
  const affectedAnalysis = state[analysisId]
  let updatedAnalysis = update(affectedAnalysis, {
    trips: { $set: trips }
  })

  // calculate array of values for each metric
  const vals = {}
  const modes = Object.keys(trips[0]).filter((mode) => (['commuterId']).indexOf(mode) === -1)
  modes.forEach((mode) => {
    vals[mode] = {
      cost: [],
      distance: [],
      time: []
    }
  })
  trips.forEach((trip) =>
    modes.forEach((mode) =>
      METRICS.forEach((metric) => {
        vals[mode][metric].push(trip[mode][metric])
      })
    )
  )

  // sort arrays
  modes.forEach((mode) =>
    METRICS.forEach((metric) => {
      vals[mode][metric].sort()
    })
  )

  updatedAnalysis = update(updatedAnalysis, {
    tripVals: { $set: vals }
  })

  // calculate analysis summary metrics
  // Potential Savings is an overall average of the difference in cost between
  // the most likely trip and the driving trip
  const len = trips.length
  let totalCostDifference = 0
  let totalTravelTime = 0
  let totalDistance = 0
  trips.forEach((trip) => {
    totalCostDifference += trip.car.cost - trip.mostLikely.cost
    totalTravelTime += trip.mostLikely.time
    totalDistance += trip.mostLikely.distance
  })

  updatedAnalysis = update(updatedAnalysis, {
    summary: {
      $set: {
        avgTravelTime: fixedRound(totalTravelTime / len),
        avgDistance: fixedRound(totalDistance / len),
        savingsPerTrip: fixedRound(totalCostDifference / len),
        savingsPerTripYear: fixedRound(totalCostDifference / len * 260),
        savingsTotalPerDay: fixedRound(totalCostDifference),
        savingsTotalPerYear: fixedRound(totalCostDifference * 260)
      }
    }
  })

  return addToEntityMap(state, updatedAnalysis)
}

export const initialState = {}
