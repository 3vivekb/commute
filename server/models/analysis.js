const {Schema} = require('mongoose')

const db = require('../db')
const Commuter = db.model('Commuter', require('./commuter'))
const Group = db.model('Group', require('./group'))
const trashPlugin = require('./plugins/trash')
const Site = db.model('Site', require('./site'))
const Trip = db.model('Trip', require('./trip'))
const dbUtils = require('../utils/db')
const later = require('../utils/later')

const schema = new Schema({
  calculationStatus: {
    default: 'calculating',
    required: true,
    type: String
  },
  groupId: {
    ref: 'Group',
    required: true,
    type: Schema.Types.ObjectId
  },
  name: {
    required: true,
    type: String
  },
  numCommuters: {
    default: 0,
    required: true,
    type: Number
  },
  organizationId: {
    ref: 'Organization',
    required: true,
    type: Schema.Types.ObjectId
  },
  siteId: {
    ref: 'Site',
    required: true,
    type: Schema.Types.ObjectId
  },
  summary: {
    // fields
    avgTravelTime: Number,
    avgDistance: Number,
    savingsPerTrip: Number,
    savingsPerTripYear: Number,
    savingsTotalPerDay: Number,
    savingsTotalPerYear: Number,

    // mongoose
    default: {
      avgTravelTime: 0,
      avgDistance: 0,
      savingsPerTrip: 0,
      savingsPerTripYear: 0,
      savingsTotalPerDay: 0,
      savingsTotalPerYear: 0
    },
    required: true,
    type: Schema.Types.Mixed
  }
})

schema.plugin(trashPlugin)

schema.pre('save', dbUtils.makeCascadeDeleteModelsFn('analysisId', [Trip]))

schema.pre('save', true, function (next, done) {
  next()

  const self = this

  if (['skipCalculation', 'calculated'].indexOf(this.calculationStatus) !== -1 ||
    this.trashed) return done()

  const profiler = require('../utils/profiler')

  // initiate profile analysis of all commuters in group
  // find all commuters via groupId
  // make sure group isn't trashed
  const ensureGroupExists = Group.count({ _id: this.groupId, trashed: undefined }).exec()
  const findCommuters = Commuter.find({ groupId: this.groupId, trashed: undefined }).exec()
  const getSite = Site.findOne({ _id: this.siteId, trashed: undefined }).exec()
  Promise.all([findCommuters, getSite, ensureGroupExists])
    .then((results) => {
      // verify that group, commuters and site exist
      const commuters = results[0]
      const site = results[1]
      const groupCount = results[2]

      if (commuters.length === 0) return done('No commuters found!')
      if (!site) return done('Site not found!')
      if (groupCount === 0) return done('Group not found!')

      self.numCommuters = commuters.length

      later(() => {
        profiler({
          analysisId: self._id,
          commuters,
          site
        })
      })
      done()
    })
    .catch((err) => {
      console.error(err)
      self.calculationStatus = 'error'
      done()
    })
})

module.exports = schema
