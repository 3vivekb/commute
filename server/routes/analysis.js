const models = require('../models')
const makeRestEndpoints = require('../utils/restEndpoints')

module.exports = function makeRoutes (app) {
  makeRestEndpoints(app,
    {
      childModels: [{
        foreignKey: 'anlaysisId',
        key: 'trips',
        model: models.Trip
      }],
      commands: {
        'Collection GET': {},
        'Collection POST': {},
        'GET': {},
        'DELETE': {}
      },
      model: models.Analysis,
      name: 'analysis'
    }
  )
}
