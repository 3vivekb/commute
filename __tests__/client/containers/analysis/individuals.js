/* global describe, expect, it */

import {mount} from 'enzyme'
import {mountToJson} from 'enzyme-to-json'
import pretty from 'pretty'
import React from 'react'
import {Provider} from 'react-redux'

import {makeMockStore, mockStores} from '../../../test-utils/mock-data.js'
import '../../../test-utils/mock-leaflet'

import {Individuals} from '../../../../client/containers/analysis'

const mockStore = makeMockStore(mockStores.withAnalysisRun)

describe('Container > Analysis > Individuals', () => {
  it('Individuals View loads (base case)', () => {
    // mount component
    const tree = mount(
      <Provider store={mockStore}>
        <Individuals
          params={{analysisId: 'analysis-2'}}
          />
      </Provider>
      , {
        attachTo: document.getElementById('test')
      }
    )
    expect(mountToJson(tree.find('.individuals-header'))).toMatchSnapshot()
    expect(pretty(tree.find('.individuals-content').html())).toMatchSnapshot()
  })

  /* it('Select commuter', () => {
    // Given a logged-in user is viewing the Individual Commuter Analysis View
    // When the user selects a row of a commuter in the table
    // Then the CarFreeAtoZ map and sidebar should get updated with the commute possibilities for the selected commuter
  }) */
})
