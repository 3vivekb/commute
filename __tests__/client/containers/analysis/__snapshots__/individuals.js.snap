exports[`Container > Analysis > Individuals Individuals View loads (base case) 1`] = `
<Connect(Individuals)
  params={
    Object {
      "analysisId": "1",
      "organizationId": "2",
    }
  }>
  <Individuals
    analysis={
      Object {
        "groupId": "1",
        "id": "1",
        "lastRunDateTime": 1477697490,
        "name": "An Analysis",
        "siteId": "1",
        "summary": Object {
          "avgDistance": 12.34,
          "avgTravelTime": 1234,
          "savingsPerTrip": 1.23,
          "savingsPerYear": 12345,
        },
        "tripVals": Object {
          "bike": Object {
            "cost": Array [
              0.12,
            ],
            "distance": Array [
              23,
            ],
            "time": Array [
              2345,
            ],
          },
          "car": Object {
            "cost": Array [
              6.78,
            ],
            "distance": Array [
              18,
            ],
            "time": Array [
              1234,
            ],
          },
          "transit": Object {
            "cost": Array [
              3.45,
            ],
            "distance": Array [
              30,
            ],
            "time": Array [
              3456,
            ],
          },
          "walk": Object {
            "cost": Array [
              0,
            ],
            "distance": Array [
              19,
            ],
            "time": Array [
              6789,
            ],
          },
        },
        "trips": Array [
          Object {
            "bike": Object {
              "cost": 0.12,
              "distance": 23,
              "polygon": "encoded",
              "time": 2345,
            },
            "car": Object {
              "cost": 6.78,
              "distance": 18,
              "polygon": "encoded",
              "time": 1234,
            },
            "commuterId": "1",
            "transit": Object {
              "cost": 3.45,
              "distance": 30,
              "polygon": "encoded",
              "time": 3456,
            },
            "walk": Object {
              "cost": 0,
              "distance": 19,
              "polygon": "encoded",
              "time": 6789,
            },
          },
        ],
      }
    }
    dispatch={[Function]}
    groupName="Mock Group"
    organizationId="2"
    params={
      Object {
        "analysisId": "1",
        "organizationId": "2",
      }
    }
    siteName="Acme Corp">
    <h1>
      Analysis
    </h1>
  </Individuals>
</Connect(Individuals)>
`;
